### PR TITLE
[PD-2120]  Count communications records per month.

### PR DESCRIPTION
**Pull request by darrint**
To test run manage.py reset_dynamic_data.
Go to https://secure.my.jobs/reports/view/dynamicoverview/#
Pick partners, then pick Number of Communication Records...

### CODE REVIEW

File: myreports/datasources/base.py [hunk 1]
```diff
@@ -107,21 +108,3 @@ def filter_query_set(self, query_set):
     it should do extra parsing on the json data for that field.
     """
     __metaclass__ = ABCMeta
-
-    @abstractmethod
-    def filter_query_set(self, query_set):
-        """Filter the given query_set.
```
Comment (darrint) [inline myreports/datasources/base.py]: The use of this method is internal to the datasource. With this more complex query I found that this single method api wasn't sufficient. I decided that this detail should be treated as internal to the datasource. I'm only documenting here what is needed for the filter to work with DataSourceJsonDriver.

File: myreports/datasources/partners.py [hunk 2]
```diff
@@ -1,41 +1,129 @@
...
+        qs_optimized = qs_ordered.prefetch_related(
+            'tags', 'primary_contact')
+
+        partners = from_django(self.partner_row_builder, qs_optimized)
+        return list(partners)
+
+    def run_count_comm_rec_per_month(self, company, filter_spec, order):
+        # Get a list of valid partners (according to our filter).
+        qs_filtered = filter_spec.filter_partners(company)
+        partner_ids = [r['id'] for r in qs_filtered.values('id').all()]
```
Comment (tdphillips) [inline myreports/datasources/partners.py]: would `.values_list('id', flat=True)` be helpful at all here?

File: myreports/datasources/comm_records.py [hunk 3]
```diff
@@ -15,7 +16,11 @@
 
 
 class CommRecordsDataSource(DataSource):
-    def run(self, company, filter_spec, order):
+    def run(self, data_type, company, filter_spec, order):
```
Comment (darrint) [inline myreports/datasources/comm_records.py]: All the datasources need to include the data_type parameter in their run method.

File: myreports/datasources/contacts.py [hunk 4]
```diff
@@ -13,7 +14,11 @@
 
 
 class ContactsDataSource(DataSource):
-    def run(self, company, filter_spec, order):
+    def run(self, data_type, company, filter_spec, order):
```
Comment (darrint) [inline myreports/datasources/contacts.py]: All the datasources need to include the data_type parameter in their run method.

File: myreports/datasources/partners.py [hunk 5]
```diff
@@ -1,41 +1,129 @@
...
-            return None
-        else:
-            return contact.name
+        qs_optimized = qs_ordered.prefetch_related(
+            'tags', 'primary_contact')
+
+        partners = from_django(self.partner_row_builder, qs_optimized)
+        return list(partners)
+
+    def run_count_comm_rec_per_month(self, company, filter_spec, order):
```
Comment (darrint) [inline myreports/datasources/partners.py]: See below.

MySql can do some basic counting. Allow that. 

Django ORM does a nice job of filtering. Continue to use that.

Pull in the data we need and merge it together in Python.

Use the RowStream mechanism to help with this.

File: myreports/datasources/partners.py [hunk 6]
```diff
@@ -105,17 +193,6 @@ def help_data_source(self, company, filter_spec, partial):
             for c in data_sources_qs
         ]
 
-    def filtered_query_set(self, company, filter_spec):
```
Comment (darrint) [inline myreports/datasources/partners.py]: Move this code down to the filter class.

File: myreports/datasources/sql.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    join
+    mypartners_contact contact_table
+    on partner_table.id = contact_table.partner_id
+    join
+    mypartners_contactrecord comm_rec_table
+    on contact_table.id = comm_rec_table.contact_id
+where
+    partner_table.id in %(partner_list)s
+group by partner_table.id, year, month
+'''
```
Comment (darrint) [inline myreports/datasources/sql.py]: MySql can at least do this. Let it.

File: myreports/datasources/util.py [hunk 8]
```diff
@@ -18,6 +18,23 @@ def dispatch_help_by_field_name(ds, company, filter_spec, field, partial):
     return method(company, filter_spec, partial)
 
 
+def dispatch_run_by_data_type(ds, data_type, company, filter_spec, order_spec):
```
Comment (darrint) [inline myreports/datasources/util.py]: This utility is an optional way to dispatch different data types to different run_* methods on datasource classes.

File: myreports/tests/setup.py [hunk 9]
```diff
@@ -495,6 +504,86 @@ def create_full_fixture():
         multi_value_expansion=False,
         has_help=True)
 
+    ConfigurationColumnFactory.create(
```
Comment (darrint) [inline myreports/tests/setup.py]: Create a new configuration to expose the new data type to the UI.

File: myreports/datasources/partners.py [hunk 10]
```diff
@@ -1,41 +1,129 @@
...
+        db_counts = {}
+        for record in count_stream:
+            partner_id = record['partner_id']
+            year = record['year']
+            month = record['month']
+            comm_rec_count = record['comm_rec_count']
+
+            if year < min_year or min_year is None:
+                min_year = year
+            if year < max_year or max_year is None:
```
Comment (tdphillips) [inline myreports/datasources/partners.py]: Should this comparison be reversed (year > max_year)?

File: myreports/tests/test_partners_datasource.py [hunk 11]
```diff
@@ -99,15 +100,120 @@ def setUp(self):
     def test_run_unfiltered(self):
         """Make sure we only get data for this user."""
         ds = PartnersDataSource()
-        recs = ds.run(self.company, PartnersFilter(), [])
+        recs = ds.run_unaggregated(self.company, PartnersFilter(), [])
         names = set([r['name'] for r in recs])
         expected = {self.partner_a.name, self.partner_b.name}
         self.assertEqual(expected, names)
+        expected_tags = {
```
Comment (darrint) [inline myreports/tests/test_partners_datasource.py]: Just testing more of the partners data here.

File: myreports/tests/test_partners_datasource.py [hunk 12]
```diff
@@ -99,15 +100,120 @@ def setUp(self):
...
+        }
+
+        def tags_by_pk(rec):
+            return (rec['partner_id'], set(t['name'] for t in rec['tags']))
+
+        found_tags = dict(tags_by_pk(r) for r in recs)
+
+        self.assertEqual(expected_tags, found_tags)
+
+    def test_run_count_comm_rec_per_month(self):
```
Comment (darrint) [inline myreports/tests/test_partners_datasource.py]: Sending RawSQL to MySQL with the bound "in" where clause can be tricky. These tests are needed now.

File: myreports/tests/test_partners_datasource.py [hunk 13]
```diff
@@ -99,15 +100,120 @@ def setUp(self):
...
+            (self.partner_b.name, 2015, 6, 0),
+            (self.partner_b.name, 2015, 5, 0),
+            (self.partner_b.name, 2015, 4, 3),
+            (self.partner_b.name, 2015, 3, 3),
+            (self.partner_b.name, 2015, 2, 3),
+            (self.partner_b.name, 2015, 1, 0),
+        ]
+        self.assertEqual(expected, data)
+
+    def test_run_count_comm_rec_per_month_one_partner(self):
```
Comment (darrint) [inline myreports/tests/test_partners_datasource.py]: Sending RawSQL to MySQL with the bound "in" where clause can be tricky. These tests are needed now.

File: myreports/tests/test_views.py [hunk 14]
```diff
@@ -756,3 +760,56 @@ def test_dynamic_partners_report_xlsx(self):
         self.assertEquals(200, resp.status_code)
         self.assertIn('The_Report.xlsx', resp['content-disposition'])
         self.assertIn('application/vnd.', resp['content-type'])
+
+    def test_dynamic_partners_report_comm_per_month(self):
```
Comment (darrint) [inline myreports/tests/test_views.py]: Do an end-to-end test using this new data type.

File: requirements.txt [hunk 15]
```diff
@@ -2,7 +2,7 @@ AuthorizeSauce==0.3.1
 BeautifulSoup==3.2.1
 cssselect==0.9.1
 Django==1.6.5
-MySQL-python==1.2.3
+MySQL-python==1.2.5
```
Comment (darrint) [inline requirements.txt]: The old version had bugs with binding "in" where clauses.

File: myreports/datasources/sql.py [hunk 16]
```diff
@@ -0,0 +1,18 @@
+count_comm_rec_per_month_per_partner_sql = '''
+select
```
Comment (tdphillips) [inline myreports/datasources/sql.py]: @JLMcLaugh Did you want to check this out? Nothing stood out to me in the rest of the code.

**Pull request merged**